### PR TITLE
Add level-zero-opt tool 

### PR DESCRIPTION
**Pull request by nbpatel**
Please review these guidelines to help with the review process:
- [x] Have you provided a meaningful PR description?
- [x] Have you added a test, a reproducer, or a reference to an issue with a reproducer?
- [x] Have you tested your changes locally for CPU and GPU devices?
- [ ] Have you made sure that new changes do not introduce compiler warnings?
- [x] If this PR is a work in progress, are you filing the PR as a draft?
- [x] Have you organized your commits logically and ensured each can be built by itself?

This PR adds a directory structure to the mlir/test directory which is similar to upstream dir structure. 
It adds a tool called level-zero-opt which allows us to run individual pass from command line and also adds a few test cases for doing some basic IR tests. 
More test cases to be added.

### CODE REVIEW

File: mlir/test/Conversion/gpu-to-spirv.mlir [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    // CHECK:  %7 = spv.InBoundsPtrAccessChain %arg2[%1] : !spv.ptr<f32, CrossWorkgroup>, i64
+    // CHECK:   spv.Store "CrossWorkgroup" %7, %6 ["Aligned", 4] : f32
+    // CHECK:   spv.Return
+    // CHECK:  }
+    // CHECK: }
+
+  }
+  func.func private @fillResource1DFloat(memref<?xf32>, f32)
+  func.func private @printMemrefF32(memref<*xf32>)
+}
```
Comment (Hardcode84) [inline mlir/test/Conversion/gpu-to-spirv.mlir:58]: Newline at the end of the file

File: mlir/test/Conversion/gpu-to-spirv.mlir [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+      gpu.return
+    }
+
+    // CHECK: spv.module @__spv__main_kernel Physical64 OpenCL {
+    // CHECK: spv.GlobalVariable @__builtin_var_WorkgroupId__ built_in("WorkgroupId") : !spv.ptr<vector<3xi64>, Input>
+    // CHECK: spv.func @main_kernel(%arg0: !spv.ptr<f32, CrossWorkgroup>, %arg1: !spv.ptr<f32, CrossWorkgroup>, %arg2: !spv.ptr<f32, CrossWorkgroup>) "None" attributes {spv.entry_point_abi = #spv.entry_point_abi<>, workgroup_attributions = 0 : i64} {
+    // CHECK: spv.Branch ^bb1
+    // CHECK: ^bb1:  // pred: ^bb0
+    // CHECK: %__builtin_var_WorkgroupId___addr = spv.mlir.addressof @__builtin_var_WorkgroupId__ : !spv.ptr<vector<3xi64>, Input>
+    // CHECK: %0 = spv.Load "Input" %__builtin_var_WorkgroupId___addr : vector<3xi64>
```
Comment (Hardcode84) [inline mlir/test/Conversion/gpu-to-spirv.mlir:42]: Generated names like `%0` can be unstable between MLIR versions, please use `%[[VAR1:.*]] = spv.Load ...` `%[[VAR2:.*]] = spv.CompositeExtract %[[VAR1]][0 : i32] ` name matching style

File: mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+func.func @main() {
+  %c8 = arith.constant 8 : index
+  %c1 = arith.constant 1 : index
+  // CHECK: func.func @main()
+  %0 = memref.alloc() : memref<8xf32>
+  %1 = memref.alloc() : memref<8xf32>
+  %2 = memref.alloc() : memref<8xf32>
+  // CHECK: %memref = gpu.alloc  () : memref<8xf32>
+  // CHECK: %memref_0 = gpu.alloc  () : memref<8xf32>
+  // CHECK: %memref_1 = gpu.alloc  () : memref<8xf32>
```
Comment (Hardcode84) [inline mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir:12]: Please check there are actually loads from these gpu allocs inside gpu.launch

File: mlir/tools/level-zero-opt/main.cpp [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+#include <mlir/IR/MLIRContext.h>
+#include <mlir/InitAllDialects.h>
+#include <mlir/InitAllPasses.h>
+#include <mlir/Tools/mlir-opt/MlirOptMain.h>
+
+int main(int argc, char **argv) {
+  mlir::registerAllPasses();
+  mlir::DialectRegistry registry;
+  mlir::registerAllDialects(registry);
+  return mlir::failed(MlirOptMain(argc, argv,
```
Comment (fschlimb) [inline mlir/tools/level-zero-opt/main.cpp:24]: How do your custom dialects and passes get added/registered?

Comment (nbpatel) [reply mlir/tools/level-zero-opt/main.cpp:24]: We need this tool now for adding more test cases...we can get rid of it once we move to refactor branch

File: mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir (hunk 3 above)
Comment (nbpatel) [reply mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir:12]: what do you mean?

File: mlir/test/Conversion/gpu-to-spirv.mlir (hunk 2 above)
Comment (nbpatel) [reply mlir/test/Conversion/gpu-to-spirv.mlir]: done

File: mlir/test/Conversion/gpu-to-spirv.mlir (hunk 1 above)
Comment (nbpatel) [reply mlir/test/Conversion/gpu-to-spirv.mlir]: done

File: mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir (hunk 3 above)
Comment (Hardcode84) [reply mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir:12]: Something like
```
// CHECK: %[[MEMREF0:.*]]= gpu.alloc  () : memref<8xf32>
// CHECK: gpu.launch
// CHECK: memref.load %[[MEMREF0]][...]
// CHECK: gpu.terminator
```

File: mlir/tools/level-zero-opt/main.cpp (hunk 4 above)
Comment (fschlimb) [reply mlir/tools/level-zero-opt/main.cpp:24]: What I am trying to say is that the refactor branch already has imex-opt (which is the equivalent of level0-opt) and the infrastructure is ready to actually run your tests in CI. What is missing in refactor that keeps you from issuing the PR against refactor?

**Comment by Hardcode84** [comment]
@nbpatel speaking of enabling/disabling individual lit tests, please take a look https://llvm.org/docs/TestingGuide.html#constraining-test-execution

File: mlir/tools/level-zero-opt/main.cpp (hunk 4 above)
Comment (fschlimb) [reply mlir/tools/level-zero-opt/main.cpp:24]: @Hardcode84 @nbpatel I still do not understand how the MLIR passes get registered. Can you enlighten me?

Comment (fschlimb) [reply mlir/tools/level-zero-opt/main.cpp:24]: Ok, so it is indeed the static vars. Very weird. Is this the recommend way of doing this?

File: mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir (hunk 3 above)
Comment (Hardcode84) [reply mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir:12]: @nbpatel please address this comment and then it is LGTM

Comment (nbpatel) [reply mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir:12]: @lchang552

Comment (lchang552) [reply mlir/test/Transforms/memref-alloc-to-gpu-alloc.mlir]: This is revised. Please let me know if further changes are required. -Liangliang

**Review by Hardcode84 — APPROVED** [review]

**Review by fschlimb — CHANGES_REQUESTED** [review]
Please make sure tests are actually run (and succeed) in CI.
You might want to look at refactor branch.

**Comment by lchang552** [comment]
> Please make sure tests are actually run (and succeed) in CI. You might want to look at refactor branch.
Run cd _build || exit 1
[0/1] cd /home/runner/work/mlir-extensions/mlir-extensions/_build/mlir/test && /usr/share/miniconda/bin/python3.9 /usr/share/miniconda/bin/lit -sv /home/runner/work/mlir-extensions/mlir-extensions/_build/mlir/test
-- Testing: 5 tests, 2 workers --
Testing:  0.. [10](https://github.com/intel/mlir-extensions/runs/7585907215?check_suite_focus=true#step:15:11).. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90.. 

Testing Time: 0.31s
  Unsupported: 3
  Passed     : 2


Hi, The tests can run correctly if you check the output of the CI for the latest commit. Three are unsupported since they are end-to-end tests and the CI is not with the target GPU.

**Comment by fschlimb** [comment]
> Three are unsupported since they are end-to-end tests and the CI is not with the target GPU.

->the GPU-related code is largely not covered by CI. This should be fixed.

**Comment by fschlimb** [comment]
ok, I see now, my bad. I had missed the "mlir" in the path.

Am I right that in the CI
* the tests in <root>/test are never run
* l0-runner tests are never run
* all tests for l0-opt are run and pass

?

**Comment by lchang552** [comment]
Hi, yes, you are right. This commit currently only enables tests in mlir/test/Conversion and mlir/test/Transform, others are disabled. The numbers reported are only for tests in mlir/test.

**Pull request merged**